Todos with Reflux/JSData

natatat opened the pull request:
Fixes https://github.com/looker/helltool/issues/14419

@aminer-looker check it out!

natatat committed:
Add jsdata/reflux dependencies
+ ReadOnlyView
+ EVENT constant

natatat committed:
Add todo module with TodoItem, TodoList, & TextInput (Stores/Actions/Directives/Controllers)
- require them in client.coffee

natatat committed:
Update index.jade and remove sample module

natatat committed:
Update readme

natatat committed:
Add JSData todo model/schema

natatat commented on src/coffee/client/modules/todo/todo_item_module.coffee:
```diff
@@ -0,0 +1,69 @@
+            @trigger(EVENT.CHANGE)
+
+        onDoItem: (todo, field, value, commit) ->
+            # todo._actions.done(!todo.done)
```
I ultimately still couldn't figure out how to call the `commit` function here. How do I access the correct `commit` fn to pass to `TodoItemActions.doItem()` (which is called [here](https://github.com/natatat/dataflux-todos/pull/1/files#diff-ba8c9981701fbdd5eb65d6d91df6e85cR54))?

aminer-looker replied:
The final argument _is_ the `commit` function you're looking for.  Just call it with the value you'd like to use: `commit(value)`

natatat replied:
But do I somehow have to pass these arguments to `TodoItemActions.doItem` when I call it? Like: `TodoItemActions.doItem(todo, field, value, commit)`?

Logging `commit` here shows that it's `undefined`.

aminer-looker replied:
Oh... sorry... I missed it before.  The action is handled by the `ReadOnlyView`, and therefore, the store doesn't need a handler of its own.  Instead, it should have a `_validate<action>` method which will be called by the `ReadOnlyView`'s handler.  _That's_ the method which will be passed the `commit` function.

natatat commented on src/coffee/client/modules/todo/todo_item_module.coffee:
```diff
@@ -0,0 +1,69 @@
+            @_items = []
+
+        onAddItem: -> # (2)
+            @_todo = new ReadOnlyView({
```
The DSLocalStorageAdapter now works, but I haven't yet configured the actual JSData implementation. So at the moment, todos still do not persist.

I'm curious about the relationship between `TodoItemStore` and `TodoListStore` – it seems like the `TodoItemStore` is doing too much and handling most of the "List" logic.

aminer-looker replied:
I think that's mostly a result of this app being very simple.  If you had multiple users, the item store might contain all items for all users, while the list store might keep track of just the items belonging to the current user.  As things stand, you could probably dispense with the list store altogether.

aminer-looker replied:
On an unrelated note... why is this being kept in an instance variable?  The way it's being used here makes it look like a local variable, and you don't seem to use it outside this context.

aminer-looker commented on src/coffee/client/client.coffee:
```diff
@@ -3,22 +3,36 @@
 
+# Load JSData libraries. The core library must be first.
+require 'js-data'
+window.DSLocalStorageAdapter = require 'js-data-localstorage'
```
For something which looks odd (as this does), it's usually best to add a comment explaining why it has to be this way.  In actual production code, it's probable that some other developer will come along and "clean up" this strange piece of code and break things.  Leaving a comment will prevent that.

aminer-looker commented on src/coffee/client/client.coffee:
```diff
@@ -3,22 +3,36 @@
+require './modules/todo/todo_item_module'
+require './modules/todo/todo_list_module'
+require './modules/todo/input_module'
+require './../read_only_view'
```
Minor point, but the `ReadOnlyView` is a client-side thing, and therefore could have been included in the `client` directory.

aminer-looker commented on src/coffee/client/modules/client_schema.coffee:
```diff
@@ -0,0 +1,18 @@
+schema.factory 'Todo', (DS) -> DS.defineResource require '../../model/todo_model'
+
+# force the services to be created so that recursive lookups succeed
+schema.run (Todo) -> # do nothing
```
Here's an example of an "odd" piece of code I documented so that someone doesn't  "clean it up" later.

aminer-looker commented on src/coffee/client/modules/todo/input_module.coffee:
```diff
@@ -0,0 +1,53 @@
+1. User types in the input, `$scope.setValue(value)` is called with ngChange, which fires TextInput's setValue action
+2. `onSetValue()` runs and updates the TextInputStore's value
+3. TextInputStore sends a notification that something has changed
+4. TextInputController listens to TextInputStore's notifications & updates it's $scope with the new TextInputStore value
```
Nice!

aminer-looker commented on src/coffee/client/modules/todo/input_module.coffee:
```diff
@@ -0,0 +1,53 @@
+        )
+
+        $scope.setValue = (value) ->
+            TextInputActions.setValue(value) # (1)
```
Interesting.  In the main codebase, I've always separated the controllers from the directives, but I think I might like this pattern even better.  It makes for one less file in the file system, and it prevents the controller from being used apart from its matching directive.  Both positive changes over the current pattern.  :+1:

aminer-looker commented on src/coffee/client/modules/todo/input_module.coffee:
```diff
@@ -0,0 +1,53 @@
+    controllerAs: 'controller'
+    controller: ($scope, TextInputActions, TextInputStore) ->
+        TextInputStore.$listen($scope, (event, id) ->
+            $scope.value = TextInputStore._value # (4)
```
You should never access a private variable from another class (the underscore is to remind you which are private and which are public).  Instead, your store should provide a `getValue` method which is used in the controller or any other class.

aminer-looker commented on src/coffee/client/modules/todo/todo_item_module.coffee:
```diff
@@ -0,0 +1,69 @@
+        onAddItem: -> # (2)
+            @_todo = new ReadOnlyView({
+                id: _.uniqueId 'todo-'
+                description: TextInputStore._value
```
Don't use another object's private variables!  See comment above.

aminer-looker commented on src/coffee/client/modules/todo/todo_item_module.coffee:
```diff
@@ -0,0 +1,69 @@
+        onRemoveItem: (id) ->
+            @_items = _.filter @_items, (item) -> item.id isnt id
+
+            @trigger(EVENT.REMOVE, @_todo.id)
```
`@_todo` is going to reference the most recently created item, not necessarily the one being removed.  You should use `id` in the trigger statement as it's _that_ one which was actually removed.

aminer-looker commented on src/coffee/client/modules/todo/todo_item_module.coffee:
```diff
@@ -0,0 +1,69 @@
+            @trigger(EVENT.CHANGE) # (4)
+
+        onRemoveItem: (id) ->
+            @_items = _.filter @_items, (item) -> item.id isnt id
```
An alternative implementation is to use a loop and the `splice` method:

``` coffeescript
for item, index in @_items
    continue unless item.id is id
    @_items.splice index, 1
    break
```

The major difference is that the current code will create a brand new array while the alternative does not (but still needs to re-index the array).  Depending upon the circumstances, you might need one instead of the other.  My default is to use `splice` as it avoids creating new objects and giving the garbage collector more work.

aminer-looker commented on src/coffee/client/modules/todo/todo_item_module.coffee:
```diff
@@ -0,0 +1,69 @@
+
+
+todo.factory 'TodoItemActions', (reflux) ->
+    reflux.createActions ['addItem', 'removeItem', 'doItem']
```
I'm obnoxiously compulsive about keeping things in alphabetical order.  These sorts of lists have a way of growing over time, and while people will be careful to preserve some sort of "logical" ordering at first, over time it breaks down and becomes a confusing mess.  Better to use a less "logical" sorting order which will always remain obvious to new developers forever.

aminer-looker commented on src/coffee/client/modules/todo/todo_list_module.coffee:
```diff
@@ -0,0 +1,54 @@
+        listenables: TodoListActions
+
+        onLoadAll: ->
+            @_todos = TodoItemStore._items
```
Private!

aminer-looker commented on src/coffee/client/modules/todo/todo_list_module.coffee:
```diff
@@ -0,0 +1,54 @@
+
+todo = angular.module('todo')
+
+todo.factory 'TodoListStore', (Todo, TodoItemStore, TodoListActions, ReadOnlyView, reflux) ->
```
I think you might do better to just remove this store altogether and just add a `getAll` method on your item store.

aminer-looker commented on src/coffee/client/modules/todo/todo_list_module.coffee:
```diff
@@ -0,0 +1,54 @@
+        )
+
+        TodoListStore.$listen($scope, (event, id) ->
+            $scope.todos = TodoListStore._todos  # (1)
```
Private!

aminer-looker commented on src/coffee/client/modules/todo/todo_list_module.coffee:
```diff
@@ -0,0 +1,54 @@
+            $scope.todos = TodoListStore._todos  # (1)
+        )
+
+        $scope.removeTodoItem = (id) -> # (3)
```
I think it would be better to have each to-do list item be it's own directive.  Each could accept a scope-bound attribute for its `id`, and its controller could then add a `remove` and `makeDone` method which can just look on the scope for the correct `id`.

aminer-looker commented on src/jade/templates/todo-list.jade:
```diff
@@ -0,0 +1,10 @@
+div(ng-repeat="todo in todos")
+  input(type="checkbox" ng-click="makeDone(todo)")
+  span(ng-class="{'done': todo.done}") {{todo.description}}
+  a(href="#" ng-click="removeTodoItem(todo.id)") remove
```
I think this would be better composed as:

``` jade
div(ng-repeat="todo in todos")
    todo-item(id="{{todo.id}}")
```

Then, you can move the contents of the loop into their own directive.

aminer-looker replied:
``` coffeescript
angular.module('todo').directive 'todoItem'
    restrict: 'E'
    scope:
        id: '@'
    controller: ($scope, TodoItemActions, TodoItemStore)->
        TodoItemStore.$listen (event, id)->
            return unless id is $scope.id
            $scope.description = TodoItemStore.get($scope.id).description
```

aminer-looker commented:
I've left a bunch of comments... happy to chat face-to-face if anything is unclear!

natatat committed:
Document this weird piece of code properly

natatat committed:
Alphabetical argument ordering

natatat committed:
Add a getValue fn, don't access private vars

natatat committed:
Change to local var, pass the correct id

natatat committed:
Remove TodoList directive, access todos from getAll() in TodoItemStore

natatat committed:
Rename to text_input

natatat committed:
Add a TodoItem directive
- handles removing/doing todos
- just pass around todo.id via $scope.id
- use TodoItemStore.get() function

natatat committed:
TodoItem template

natatat commented:
@aminer-looker made changes according to your feedback! my only question is here: https://github.com/natatat/dataflux-todos/pull/1#discussion_r48445428

natatat committed:
Toggle done correctly!
